### PR TITLE
 feat: Runtime Bundle Mapping from global or window object

### PR DESCRIPTION
**Pull request by DitlevHB**
🐢 Finally ready with a pull request. 
I hope it goes through! *fingers crossed*

### CODE REVIEW

**Comment by nchanged** [comment]
@DitlevHB some tests have failed. I just want to ask you to write a few test cases
in this file

https://github.com/fuse-box/fuse-box/blob/a13815949a16c6a37a41d2264470732149d43b9e/src/tests/dynamic_imports_test/CodeSplittingFileIntegrity.test.ts#L29-L64

It's pretty simple, we need browser server and universal to check if the data is set correctly.

**Comment by DitlevHB** [comment]
Hmm, not sure how it should check if the data is set correctly.
This is what i got so far:
```
 "Should if check browser data is it correctly"() {
        return FuseTestEnv.create(
            {
                project: {
                    files: {
                        "index.ts": `
                            export async function getRemoteFile() {
                                await import("./components/HomeComponent")
                            }
                        `,
                        "components/HomeComponent.ts": `
                            export function home(){ return "home"  }
                        `,

                    },
                    plugins: [
                        QuantumPlugin({
                            target: "browser",
                            runtimeBundleMapping: "newBundleMapping"
                        })
                    ]
                }
            }
        )
            .simple().then(test => test.browser((window, env) => {
                window['newBundleMapping'] = {
                    c: {
                        b: './asdasd',
                        s: './asdasdasd'},
                    i: {}
                };

                window.$fsx.r(0);

                
                env.scriptShouldExist(HOME_COMPONENT_SCRIPT);
               
                const appScript = env.getScript("app.js");
                
                appScript.shouldNotFindString('// default/components/HomeComponent.js');
                

                const homeScript = env.getScript(HOME_COMPONENT_SCRIPT)
                homeScript.shouldFindString('// default/components/HomeComponent.js');
            }));
    }

```

Wouldn't it make sense to try and import the homecomponent and see if it is successfull?

**Comment by nchanged** [comment]
` window.$fsx.r(0);` that's the import of a first module

**Comment by nchanged** [comment]
You can run a single test
`node test --file=CodeSplittingFileIntegrity.test.ts`

File: src/tests/dynamic_imports_test/CodeSplittingFileIntegrity.test.ts [hunk 1]
```diff
@@ -705,6 +706,58 @@ export class CodeSplittingFileIntegrityTest {
...
+                            runtimeBundleMapping: "newBundleMapping"
+                        })
+                    ]
+                }
+            }
+        )
+            .simple().then(test => test.browser((window, env) => {
+                window['newBundleMapping'] = {
+                    c: {
+                        b: './asdasd',
```
Comment (nchanged) [inline src/tests/dynamic_imports_test/CodeSplittingFileIntegrity.test.ts]: Let's give it real values, and see if it's possible to import without errors

Comment (DitlevHB) [reply src/tests/dynamic_imports_test/CodeSplittingFileIntegrity.test.ts]: Well it doesn't fail with these 'non-real' values.
I wanted to make it fail first and then correct it with real values.

Comment (DitlevHB) [reply src/tests/dynamic_imports_test/CodeSplittingFileIntegrity.test.ts]: It should fail on 
````
window.$fsx.r(0);
```
right?
Or am i missing the point

Comment (nchanged) [reply src/tests/dynamic_imports_test/CodeSplittingFileIntegrity.test.ts]: can you hack into `FuseTestEnv` and add this `preloadScript` as an optional argument? That should do the trick

Comment (DitlevHB) [reply src/tests/dynamic_imports_test/CodeSplittingFileIntegrity.test.ts]: So i successfully extended FuseTestEnv and added the preloading script.
I'm having issues validating, that the dynamic loading works properly.

```
 .simple().then(test => test.browser((window, env) => {
                var index = window.$fsx.r(0);
                index.getRemoteFile().then(result => {
                    should(result).equal("home");
                });
            }, bundleMappingScript));
```

Seems like async operations aren't possible within the first argument of FuseTestEnv.browser?

**Comment by nchanged** [comment]
You need to return it ) if course

File: src/tests/dynamic_imports_test/CodeSplittingFileIntegrity.test.ts (hunk 1 above)
Comment (nchanged) [reply src/tests/dynamic_imports_test/CodeSplittingFileIntegrity.test.ts]: How is it going with it? Do you need help? You can just return it
```
 var index = window.$fsx.r(0);
                return index.getRemoteFile().then(result => {
                    should(result).equal("home");
                });
```

I can help you with this. Let's merge it already ;-)

Comment (DitlevHB) [reply src/tests/dynamic_imports_test/CodeSplittingFileIntegrity.test.ts]: Hey I am really busy this week, I'll see what i can do. I might not be able to check in before friday.

**Comment by DitlevHB** [comment]
Hey, have you seen I pushed the new tests?

**Comment by nchanged** [comment]
@DitlevHB one thing we forgot to add is documentation ;-(

**Comment by DitlevHB** [comment]
I'm busy with exams at the moment, but I'll add some documentation within a week or two 😉

**Pull request merged**